### ISSUE TITLE
Update to golang 1.16

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#29693]
TODO:
* [ ] Bump binary in docker image
* [x] ensure all tests pass
* [x] Ensure benchmark is reasonable
* [ ] Make use of new features
   * [ ]  https://tip.golang.org/doc/go1.16#library-embed

**Comment by howardjohn** [comment]
agent binary size decrease 38mb -> 34 mb

**Comment by howardjohn** [comment]
```
--- FAIL: TestConvertResources (0.15s)
    --- FAIL: TestConvertResources/tls (0.01s)
        conversion_test.go:48: 1 error occurred:
                * Gateway/gateway/default: spec.listeners.tls.routeOverride.certificate: Invalid value: "null": spec.listeners.tls.routeOverride.certificate in body must be of type string: "null"


FAIL

```

single unit test failure, I think its just relying on implementation details of some error messages. Should be an easy fix.

**Comment by howardjohn** [comment]
```
name                                   old time/op        new time/op        delta
RouteGeneration/gateways-6                   47.5ms ± 1%        43.5ms ± 1%   -8.50%  (p=0.008 n=5+5)
RouteGeneration/gateways-shared-6             193ms ± 2%         176ms ± 1%   -8.65%  (p=0.008 n=5+5)
RouteGeneration/empty-6                       921µs ± 2%         870µs ± 2%   -5.54%  (p=0.008 n=5+5)
RouteGeneration/tls-6                         934µs ± 4%         881µs ± 3%   -5.64%  (p=0.016 n=5+5)
RouteGeneration/telemetry-6                   929µs ± 2%         876µs ± 3%   -5.67%  (p=0.008 n=5+5)
RouteGeneration/virtualservice-6             1.94ms ± 1%        1.83ms ± 2%   -5.22%  (p=0.008 n=5+5)
ClusterGeneration/gateways-6                 11.6ms ± 1%        11.0ms ± 1%   -5.24%  (p=0.008 n=5+5)
ClusterGeneration/gateways-shared-6          11.5ms ± 1%        11.0ms ± 3%   -4.38%  (p=0.008 n=5+5)
ClusterGeneration/empty-6                    7.19ms ± 1%        6.73ms ± 2%   -6.39%  (p=0.008 n=5+5)
ClusterGeneration/tls-6                      6.99ms ± 2%        6.55ms ± 1%   -6.32%  (p=0.008 n=5+5)
ClusterGeneration/telemetry-6                4.82ms ± 1%        4.58ms ± 1%   -4.95%  (p=0.008 n=5+5)
ClusterGeneration/virtualservice-6           1.18ms ± 2%        1.13ms ± 1%   -3.95%  (p=0.008 n=5+5)
ListenerGeneration/gateways-6                17.5ms ± 2%        15.9ms ± 0%   -9.43%  (p=0.008 n=5+5)
ListenerGeneration/gateways-shared-6         34.7µs ± 2%        32.0µs ± 1%   -7.60%  (p=0.008 n=5+5)
ListenerGeneration/empty-6                   1.41ms ± 3%        1.37ms ± 1%     ~     (p=0.056 n=5+5)
ListenerGeneration/tls-6                     1.33ms ± 1%        1.32ms ± 1%     ~     (p=0.310 n=5+5)
ListenerGeneration/telemetry-6               1.52ms ± 2%        1.49ms ± 1%     ~     (p=0.056 n=5+5)
ListenerGeneration/virtualservice-6           109µs ± 1%         101µs ± 1%   -7.13%  (p=0.008 n=5+5)
NameTableGeneration/gateways-6               31.2ns ± 4%        31.5ns ± 1%     ~     (p=0.310 n=5+5)
NameTableGeneration/gateways-shared-6        31.9ns ± 5%        31.7ns ± 2%     ~     (p=0.841 n=5+5)
NameTableGeneration/empty-6                  1.38µs ± 1%        1.20µs ± 1%  -13.21%  (p=0.008 n=5+5)
NameTableGeneration/tls-6                    1.38µs ± 3%        1.18µs ± 2%  -14.80%  (p=0.008 n=5+5)
NameTableGeneration/telemetry-6              1.38µs ± 2%        1.16µs ± 3%  -15.83%  (p=0.008 n=5+5)
NameTableGeneration/virtualservice-6         1.38µs ± 3%        1.16µs ± 1%  -15.96%  (p=0.008 n=5+5)
SecretGeneration/secrets-10-6                36.4µs ± 1%        34.0µs ± 1%   -6.62%  (p=0.008 n=5+5)
SecretGeneration/secrets-1000-6              3.83ms ± 2%        3.42ms ± 2%  -10.67%  (p=0.008 n=5+5)
EndpointGeneration/1/100-6                    285µs ± 1%         279µs ± 1%   -1.99%  (p=0.008 n=5+5)
EndpointGeneration/10/10-6                   94.0µs ± 2%        88.5µs ± 6%   -5.90%  (p=0.008 n=5+5)
EndpointGeneration/100/10-6                   735µs ± 3%         651µs ± 5%  -11.40%  (p=0.008 n=5+5)
EndpointGeneration/1000/1-6                   681µs ± 7%         648µs ± 5%     ~     (p=0.095 n=5+5)

name                                   old kb/msg         new kb/msg         delta
RouteGeneration/gateways-6                    2.67k ± 0%         2.67k ± 0%     ~     (all equal)
RouteGeneration/gateways-shared-6             12.6k ± 0%         12.6k ± 0%     ~     (all equal)
RouteGeneration/empty-6                        31.8 ± 0%          31.8 ± 0%   -0.13%  (p=0.008 n=5+5)
RouteGeneration/tls-6                          31.8 ± 0%          31.8 ± 0%   -0.13%  (p=0.008 n=5+5)
RouteGeneration/telemetry-6                    31.8 ± 0%          31.8 ± 0%   -0.13%  (p=0.008 n=5+5)
RouteGeneration/virtualservice-6               80.7 ± 0%          80.7 ± 0%   +0.05%  (p=0.008 n=5+5)
ClusterGeneration/gateways-6                    293 ± 0%           293 ± 0%   +0.03%  (p=0.008 n=5+5)
ClusterGeneration/gateways-shared-6             293 ± 0%           293 ± 0%   +0.03%  (p=0.008 n=5+5)
ClusterGeneration/empty-6                       270 ± 0%           270 ± 0%   -0.11%  (p=0.008 n=5+5)
ClusterGeneration/tls-6                         265 ± 0%           265 ± 0%     ~     (all equal)
ClusterGeneration/telemetry-6                   120 ± 0%           120 ± 0%   -0.17%  (p=0.008 n=5+5)
ClusterGeneration/virtualservice-6             29.4 ± 0%          29.4 ± 0%   +0.07%  (p=0.008 n=5+5)
ListenerGeneration/gateways-6                 1.23k ± 0%         1.23k ± 0%     ~     (all equal)
ListenerGeneration/gateways-shared-6           2.17 ± 0%          2.17 ± 0%   -0.18%  (p=0.008 n=5+5)
ListenerGeneration/empty-6                     11.6 ± 0%          11.6 ± 0%   -0.43%  (p=0.008 n=5+5)
ListenerGeneration/tls-6                       8.49 ± 0%          8.49 ± 0%   +0.05%  (p=0.008 n=5+5)
ListenerGeneration/telemetry-6                 11.6 ± 0%          11.6 ± 0%   -0.43%  (p=0.008 n=5+5)
ListenerGeneration/virtualservice-6            4.96 ± 0%          4.96 ± 0%   +0.02%  (p=0.008 n=5+5)
NameTableGeneration/gateways-6                 0.00               0.00          ~     (all equal)
NameTableGeneration/gateways-shared-6          0.00               0.00          ~     (all equal)
NameTableGeneration/empty-6                    0.00               0.00          ~     (all equal)
NameTableGeneration/tls-6                      0.00               0.00          ~     (all equal)
NameTableGeneration/telemetry-6                0.00               0.00          ~     (all equal)
NameTableGeneration/virtualservice-6           0.00               0.00          ~     (all equal)
SecretGeneration/secrets-10-6                  46.0 ± 0%          46.0 ± 0%   +0.02%  (p=0.008 n=5+5)
SecretGeneration/secrets-1000-6               4.60k ± 0%         4.60k ± 0%     ~     (all equal)
EndpointGeneration/1/100-6                     5.79 ± 0%          5.79 ± 0%     ~     (all equal)
EndpointGeneration/10/10-6                     2.83 ± 0%          2.83 ± 0%     ~     (all equal)
EndpointGeneration/100/10-6                    26.2 ± 0%          26.2 ± 0%   +0.11%  (p=0.008 n=5+5)
EndpointGeneration/1000/1-6                    26.6 ± 0%          26.6 ± 0%   -0.04%  (p=0.008 n=5+5)

name                                   old resources/msg  new resources/msg  delta
RouteGeneration/gateways-6                    1.00k ± 0%         1.00k ± 0%     ~     (all equal)
RouteGeneration/gateways-shared-6              2.00 ± 0%          2.00 ± 0%     ~     (all equal)
RouteGeneration/empty-6                        2.00 ± 0%          2.00 ± 0%     ~     (all equal)
RouteGeneration/tls-6                          2.00 ± 0%          2.00 ± 0%     ~     (all equal)
RouteGeneration/telemetry-6                    2.00 ± 0%          2.00 ± 0%     ~     (all equal)
RouteGeneration/virtualservice-6               1.00 ± 0%          1.00 ± 0%     ~     (all equal)
ClusterGeneration/gateways-6                  1.00k ± 0%         1.00k ± 0%     ~     (all equal)
ClusterGeneration/gateways-shared-6           1.00k ± 0%         1.00k ± 0%     ~     (all equal)
ClusterGeneration/empty-6                       410 ± 0%           410 ± 0%     ~     (all equal)
ClusterGeneration/tls-6                         410 ± 0%           410 ± 0%     ~     (all equal)
ClusterGeneration/telemetry-6                   410 ± 0%           410 ± 0%     ~     (all equal)
ClusterGeneration/virtualservice-6              104 ± 0%           104 ± 0%     ~     (all equal)
ListenerGeneration/gateways-6                  2.00 ± 0%          2.00 ± 0%     ~     (all equal)
ListenerGeneration/gateways-shared-6           2.00 ± 0%          2.00 ± 0%     ~     (all equal)
ListenerGeneration/empty-6                     6.00 ± 0%          6.00 ± 0%     ~     (all equal)
ListenerGeneration/tls-6                       6.00 ± 0%          6.00 ± 0%     ~     (all equal)
ListenerGeneration/telemetry-6                 6.00 ± 0%          6.00 ± 0%     ~     (all equal)
ListenerGeneration/virtualservice-6            3.00 ± 0%          3.00 ± 0%     ~     (all equal)
NameTableGeneration/gateways-6                 0.00               0.00          ~     (all equal)
NameTableGeneration/gateways-shared-6          0.00               0.00          ~     (all equal)
NameTableGeneration/empty-6                    1.00 ± 0%          1.00 ± 0%     ~     (all equal)
NameTableGeneration/tls-6                      1.00 ± 0%          1.00 ± 0%     ~     (all equal)
NameTableGeneration/telemetry-6                1.00 ± 0%          1.00 ± 0%     ~     (all equal)
NameTableGeneration/virtualservice-6           1.00 ± 0%          1.00 ± 0%     ~     (all equal)
SecretGeneration/secrets-10-6                  10.0 ± 0%          10.0 ± 0%     ~     (all equal)
SecretGeneration/secrets-1000-6               1.00k ± 0%         1.00k ± 0%     ~     (all equal)
EndpointGeneration/1/100-6                      100 ± 0%           100 ± 0%     ~     (all equal)
EndpointGeneration/10/10-6                     10.0 ± 0%          10.0 ± 0%     ~     (all equal)
EndpointGeneration/100/10-6                    10.0 ± 0%          10.0 ± 0%     ~     (all equal)
EndpointGeneration/1000/1-6                    1.00 ± 0%          1.00 ± 0%     ~     (all equal)

name                                   old alloc/op       new alloc/op       delta
RouteGeneration/gateways-6                   17.0MB ± 0%        16.8MB ± 0%   -1.23%  (p=0.008 n=5+5)
RouteGeneration/gateways-shared-6            69.3MB ± 0%        68.4MB ± 0%   -1.31%  (p=0.008 n=5+5)
RouteGeneration/empty-6                       605kB ± 0%         604kB ± 0%   -0.04%  (p=0.016 n=5+4)
RouteGeneration/tls-6                         605kB ± 0%         604kB ± 0%     ~     (p=0.056 n=5+5)
RouteGeneration/telemetry-6                   606kB ± 0%         606kB ± 0%     ~     (p=0.841 n=5+5)
RouteGeneration/virtualservice-6              890kB ± 0%         884kB ± 0%   -0.76%  (p=0.008 n=5+5)
ClusterGeneration/gateways-6                 5.01MB ± 0%        4.96MB ± 1%   -1.02%  (p=0.008 n=5+5)
ClusterGeneration/gateways-shared-6          5.01MB ± 0%        4.97MB ± 0%   -0.96%  (p=0.008 n=5+5)
ClusterGeneration/empty-6                    2.79MB ± 0%        2.76MB ± 0%   -0.94%  (p=0.008 n=5+5)
ClusterGeneration/tls-6                      2.77MB ± 0%        2.75MB ± 0%   -0.71%  (p=0.008 n=5+5)
ClusterGeneration/telemetry-6                2.03MB ± 0%        2.02MB ± 0%   -0.94%  (p=0.008 n=5+5)
ClusterGeneration/virtualservice-6            505kB ± 0%         501kB ± 0%   -0.98%  (p=0.008 n=5+5)
ListenerGeneration/gateways-6                7.11MB ± 1%        7.10MB ± 0%     ~     (p=0.310 n=5+5)
ListenerGeneration/gateways-shared-6         14.9kB ± 0%        14.9kB ± 0%     ~     (p=0.421 n=5+5)
ListenerGeneration/empty-6                    744kB ± 0%         747kB ± 0%   +0.38%  (p=0.008 n=5+5)
ListenerGeneration/tls-6                      714kB ± 0%         717kB ± 0%   +0.40%  (p=0.008 n=5+5)
ListenerGeneration/telemetry-6                781kB ± 0%         783kB ± 0%   +0.31%  (p=0.008 n=5+5)
ListenerGeneration/virtualservice-6          61.4kB ± 0%        61.3kB ± 0%   -0.17%  (p=0.000 n=4+5)
NameTableGeneration/gateways-6                0.00B              0.00B          ~     (all equal)
NameTableGeneration/gateways-shared-6         0.00B              0.00B          ~     (all equal)
NameTableGeneration/empty-6                    264B ± 0%          264B ± 0%     ~     (all equal)
NameTableGeneration/tls-6                      264B ± 0%          264B ± 0%     ~     (all equal)
NameTableGeneration/telemetry-6                265B ± 0%          264B ± 0%     ~     (p=0.167 n=5+5)
NameTableGeneration/virtualservice-6           264B ± 0%          264B ± 0%     ~     (all equal)
SecretGeneration/secrets-10-6                60.3kB ± 0%        60.2kB ± 0%   -0.28%  (p=0.008 n=5+5)
SecretGeneration/secrets-1000-6              6.01MB ± 0%        6.00MB ± 0%     ~     (p=0.651 n=5+5)
EndpointGeneration/1/100-6                    101kB ± 0%          99kB ± 0%   -1.58%  (p=0.008 n=5+5)
EndpointGeneration/10/10-6                   14.7kB ± 0%        14.5kB ± 0%   -1.03%  (p=0.008 n=5+5)
EndpointGeneration/100/10-6                  63.4kB ± 0%        63.2kB ± 0%   -0.27%  (p=0.008 n=5+5)
EndpointGeneration/1000/1-6                  53.2kB ± 0%        53.1kB ± 0%     ~     (p=0.087 n=5+5)

name                                   old allocs/op      new allocs/op      delta
RouteGeneration/gateways-6                     210k ± 0%          210k ± 0%     ~     (p=0.079 n=5+5)
RouteGeneration/gateways-shared-6              907k ± 0%          908k ± 0%     ~     (p=1.000 n=5+5)
RouteGeneration/empty-6                       6.11k ± 0%         6.12k ± 0%     ~     (p=0.794 n=5+4)
RouteGeneration/tls-6                         6.12k ± 0%         6.12k ± 0%     ~     (p=0.714 n=5+5)
RouteGeneration/telemetry-6                   6.13k ± 0%         6.13k ± 0%     ~     (p=0.611 n=5+5)
RouteGeneration/virtualservice-6              11.1k ± 0%         11.1k ± 0%   -0.01%  (p=0.029 n=4+4)
ClusterGeneration/gateways-6                  81.4k ± 0%         81.4k ± 0%     ~     (p=1.000 n=5+5)
ClusterGeneration/gateways-shared-6           81.6k ± 0%         81.6k ± 0%     ~     (p=1.000 n=5+5)
ClusterGeneration/empty-6                     41.8k ± 0%         41.8k ± 0%   -0.01%  (p=0.008 n=5+5)
ClusterGeneration/tls-6                       41.4k ± 0%         41.4k ± 0%   -0.01%  (p=0.008 n=5+5)
ClusterGeneration/telemetry-6                 33.4k ± 0%         33.4k ± 0%     ~     (p=0.286 n=4+5)
ClusterGeneration/virtualservice-6            8.31k ± 0%         8.31k ± 0%     ~     (p=0.167 n=5+5)
ListenerGeneration/gateways-6                 85.4k ± 0%         85.0k ± 0%   -0.42%  (p=0.008 n=5+5)
ListenerGeneration/gateways-shared-6            215 ± 0%           214 ± 0%     ~     (p=0.095 n=5+4)
ListenerGeneration/empty-6                    14.5k ± 0%         14.2k ± 0%   -2.24%  (p=0.000 n=4+5)
ListenerGeneration/tls-6                      14.2k ± 0%         13.9k ± 0%   -2.21%  (p=0.029 n=4+4)
ListenerGeneration/telemetry-6                15.1k ± 0%         14.8k ± 0%   -2.17%  (p=0.000 n=4+5)
ListenerGeneration/virtualservice-6             831 ± 0%           824 ± 0%   -0.84%  (p=0.008 n=5+5)
NameTableGeneration/gateways-6                 0.00               0.00          ~     (all equal)
NameTableGeneration/gateways-shared-6          0.00               0.00          ~     (all equal)
NameTableGeneration/empty-6                    6.00 ± 0%          6.00 ± 0%     ~     (all equal)
NameTableGeneration/tls-6                      6.00 ± 0%          6.00 ± 0%     ~     (all equal)
NameTableGeneration/telemetry-6                6.00 ± 0%          6.00 ± 0%     ~     (all equal)
NameTableGeneration/virtualservice-6           6.00 ± 0%          6.00 ± 0%     ~     (all equal)
SecretGeneration/secrets-10-6                   189 ± 0%           189 ± 0%     ~     (all equal)
SecretGeneration/secrets-1000-6               18.0k ± 0%         18.0k ± 0%     ~     (p=0.556 n=4+5)
EndpointGeneration/1/100-6                    1.81k ± 0%         1.71k ± 0%   -5.47%  (p=0.008 n=5+5)
EndpointGeneration/10/10-6                      187 ± 0%           178 ± 0%   -4.81%  (p=0.008 n=5+5)
EndpointGeneration/100/10-6                     192 ± 0%           182 ± 0%   -5.01%  (p=0.000 n=5+4)
EndpointGeneration/1000/1-6                    25.6 ± 2%          25.0 ± 0%     ~     (p=0.095 n=5+4)
```

wow, big increase

**Comment by jacob-delgado** [comment]
ETA for golang 1.16 is 2/1. Istio release date is 2/9 with a code freeze of 2/1. If go releases late that’ll delay istio. I can see reasons for/against. I'm fine with it myself, but delaying the release if we are waiting on go could be problematic.

**Comment by howardjohn** [comment]
Agreed, I am on the fence. If we do it for 1.16, lets start updating master to the beta/rc of golang so the update in code freeze is minimal. FWIW k8s had the exact same dilemma with go 1.15 and they went for the upgrade - but I have no clue if they are happy with that decision and what different factors they have (for one, longer release cycle - I don't think go 1.15 will go out of support for us)

**Comment by jacob-delgado** [comment]
cc @shamsher31 @brian-avery @ericvn @stevenctl Thoughts? Looking at https://endoflife.date/go 1.15 will be suported until 1.17 is released (ETA August)

John, do you know anyone on the k8s team we could ask about that experience? Some of the features would be nice to get and I know there's been work in the x509 library which we use.

**Comment by ericvn** [comment]
Looking back at Istio releases, we typically move up a Golang version every 2 releases. This make sense since they release on a roughly 6 month cycle to our 3 month cycle. Continuing with that, we would release Istio 1.9 on Golang 1.15 (we released 1.8 on 1.15). What this does is makes the Istio support end roughly line up with the Golang end of support every other release. For example, Golang 1.13 (used in Istio 1.4 and 1.5) end of support would have been 8/11/20 and the Istio 1.5 end of support was 8/24/20. If we tried to move up a release so Istio 1.9 and 1.10 used Golang 1.16, we are better on the back end, but have an issue if Golang isn't released on time. And we would hit a similar problem in August with the release of Golang 1.19 and Istio 1.11.

I'm probably more comfortable leaving us on our current schedule where we are close on the back end dates and not putting a release at risk which implies we stay on Golang 1.15 for two Istio releases as we have done previously.
